### ISSUE TITLE
Best practices to install R packages?

### ISSUE THREAD

**Issue by mavam** [quarto-dev/quarto-actions#43]
I'm using quarto-actions in CI, but I'm not sure how to best prepare my R installation.

Here's what I have:

```yaml
      - name: Setup Quarto
        uses: quarto-dev/quarto-actions/setup@v2
      - name: Render Quarto Project
        uses: quarto-dev/quarto-actions/render@v2
```

This yields this error:

```
Run quarto-dev/quarto-actions/render@v2
Run if [ "" == "" ]; then
  
Error in loadNamespace(x) : there is no package called ‘jsonlite’
Calls: .main ... loadNamespace -> withRestarts -> withOneRestart -> doWithOneRestart
Execution halted

R installation:
  Version: 4.2.1
  Path: /usr/lib/R
  LibPaths:
    - /usr/local/lib/R/site-library
    - /usr/lib/R/site-library
    - /usr/lib/R/library
  rmarkdown: (None)

The rmarkdown package is not available in this R installation.
Install with install.packages("rmarkdown")

Error: Process completed with exit code 1.
```

How would go after this? Use renv?

**Comment by cderv** [comment]
We have examples and recommandation in the examples section of this  repos. See https://github.com/quarto-dev/quarto-actions/blob/main/examples/example-03-dependencies.md

You'll find also some example in our doc in the Publishing section: https://quarto.org/docs/publishing/ci.html#github-actions

Example with `quarto publish` section and executing code: https://quarto.org/docs/publishing/quarto-pub#executing-code

Hopefully, you'll find everything you need. If not, please do tell us so that we improve.

**Comment by mavam** [comment]
Thanks for the pointers! I looked through them but couldn't find recommendations on how to manage package dependencies. Specifically, I'm looking to install a subset of `tidyverse`. How would I go after it?

I tried this:
```yaml
      - name: Install tidyverse
        run: install.packages("tidyverse")
        shell: Rscript {0}
```

But I'm not getting very far:

```
ERROR: configuration failed for package ‘curl’
* removing ‘/home/runner/work/vast/vast/renv/staging/1/curl’
Traceback (most recent calls last):
12: install.packages("tidyverse")
11: install(pkgs)
10: renv_install_impl(records)
 9: renv_install_staged(records)
 8: renv_install_default(records)
 7: handler(package, renv_install_package(record))
 6: renv_install_package(record)
 5: withCallingHandlers(renv_install_package_impl(record), error = function(e) {
        vwritef("\tFAILED")
        writef(e$output)
    })
 4: renv_install_package_impl(record)
 3: r_cmd_install(package, path)
 2: r_exec_error(package, output, "install", status)
 1: stop(error)
Execution halted
Error: Process completed with exit code 1.
```

It was also an 8min installation process, which feels a bit too long.

I followed closely the docs, but it looks like I'm missing something very fundamental. I tried then this:

```
- uses: r-lib/actions/setup-r-dependencies@v2
  with:
    cache-version: 2
    extra-packages: |
      any::tidyverse
```

But it turns out I need a `DESCRIPTION` file in my top-level repo:

```
   x Cannot install packages:
  * deps::.: No root directory found in /home/runner/work/vast/vast or its parent directories. Root criterion: contains a file "DESCRIPTION" with contents matching "^Package: " 
  Execution halted
  Error: Process completed with exit code 1.
```

So now I'm a bit lost. My quarto project is in subdirectory `web/` relative to the top-level directory. I'm not sure in which direction to go forward to make it work.

**Comment by cderv** [comment]
Quarto actions currently do nothing special for the computation environment. It needs to be configured as with any other project using R. There are several solutions to install R packages on a system. Using an existing action allows to benefit from some feature. 

> ERROR: configuration failed for package ‘curl’

I think you got this error because you are missing a system dependency on your Linux environment. Probably some more steps to add if you want to do the installation manually. 

Using [pak](https://pak.r-lib.org/) to install can help with such dependency, and also https://packagemanager.rstudio.com/ to have Linux Binaries. 

But I would advice to use existing actions

> But it turns out I need a DESCRIPTION file in my top-level repo:

And yes https://github.com/r-lib/actions/tree/v2-branch/setup-r-dependencies is based on a DESCRIPTION where you would track the package you need for you project. They would then be automatically installed. Creating such file is not a bad practice at all for any project, as this way you have a file that list all the required deps. 

Also, you can look at the [Usage](https://github.com/r-lib/actions/tree/v2-branch/setup-r-dependencies#usage) section to see config parameter you can change. 

Is writing a DESCRIPTION a no go ? 

Maybe we should ask the project to remove the constraint of DESCRIPTION to use only `extra-package`. 

The solution we document in the example is also to use **renv** with `setup-renv` action. 

Anyway, thank you for your feedback. I'll remember to add a complete example of a workflow using R for computation

**Comment by mavam** [comment]
Thanks for the guidance.

> Otherwise, is writing a DESCRIPTION a no go ?
> Maybe we should ask the project to remove the constraint of DESCRIPTION more explicitly.

I went down this route ultimately. This way, it's easy to list the dependencies in the `Imports` section.

> The solution we document in the example is also to use renv with setup-renv action.

I decided against renv ultimately, because the lock file is giant and it's a bit noisy in the repo (requires adding .Rprofile, renv.lock, and renv/activate). Using pak to install the dependencies in the DESCRIPTION file is a bliss in comparison, and is also much faster.
